### PR TITLE
Issue 207

### PR DESCRIPTION
**Pull request by marijarin**
Closes https://github.com/mfvanek/pg-index-health-demo/issues/207

### CODE REVIEW

**Comment by marijarin** [comment]
I had to suppress some spot bugs warnings, is it OK?

**Comment by mfvanek** [comment]
> I had to suppress some spot bugs warnings, is it OK? If not, could you, please, give me a hint, how to meet the requirements? Especially for EI_EXPOSE_REP2

Use https://github.com/mfvanek/pg-index-health-demo/blob/master/config/spotbugs/exclude.xml
Fully suppress EI_EXPOSE_REP2 here https://github.com/mfvanek/pg-index-health-demo/blob/559bc767a11100282f0e2bbdd6c1a8299e2a8fcb/config/spotbugs/exclude.xml#L4

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationController.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    private final DbMigrationGeneratorService dbMigrationGeneratorService;
+
+    @PostMapping("/generate")
+    public ForeignKeyMigrationResponse generateFkMigration(@RequestBody final ForeignKeyMigrationRequest foreignKeyMigrationRequest) {
+        return dbMigrationGeneratorService.addIndexesWithFkChecks(foreignKeyMigrationRequest);
+    }
+
+    @ResponseStatus(HttpStatus.EXPECTATION_FAILED)
+    @org.springframework.web.bind.annotation.ExceptionHandler(RuntimeException.class)
+    public MigrationError handleMigrationException(final RuntimeException runtimeException) {
```
Comment (mfvanek) [inline pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationController.java:38]: Can we use more specific exception here?

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        final HighAvailabilityPgConnection haPgConnection = highAvailabilityPgConnectionFactory.of(credentials);
+        final DatabaseCheckOnCluster<ForeignKey> foreignKeysNotCoveredWithIndex = new ForeignKeysNotCoveredWithIndexCheckOnCluster(haPgConnection);
+        return foreignKeysNotCoveredWithIndex.check(PgContext.of("demo"));
+    }
+
+    @SuppressFBWarnings("SIL_SQL_IN_LOOP")
+    private List<String> generatedMigrations(final List<ForeignKey> foreignKeys) {
+        final List<String> generatedMigrations = dbMigrationGenerator.generate(foreignKeys);
+        log.info("Generated migrations: {}", generatedMigrations);
+        try (Connection connection = dataSource.getConnection()) {
```
Comment (mfvanek) [inline pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java:63]: Use JdbcTemplate here

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/dto/ForeignKeyMigrationRequest.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * Licensed under the Apache License 2.0
+ */
+
+package io.github.mfvanek.pg.index.health.demo.dto;
+
+import io.github.mfvanek.pg.connection.ConnectionCredentials;
+
+public record ForeignKeyMigrationRequest(
+    ConnectionCredentials credentials
```
Comment (mfvanek) [inline pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/dto/ForeignKeyMigrationRequest.java:13]: Sensitive data should not be passed as a request parameter. It is better to remove this class

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        final List<String> migrations = generatedMigrations(keysBefore);
+        final List<ForeignKey> keysAfter = getFksFromDb(fkMigrationRequest.credentials());
+        if (!keysAfter.isEmpty()) {
+            throw new IllegalStateException("There should be no foreign keys not covered by the index");
+        }
+        return new ForeignKeyMigrationResponse(keysBefore, keysAfter, migrations);
+    }
+
+    public List<ForeignKey> getFksFromDb(final ConnectionCredentials credentials) {
+        final HighAvailabilityPgConnection haPgConnection = highAvailabilityPgConnectionFactory.of(credentials);
```
Comment (mfvanek) [inline pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java:54]: We already have a bean `highAvailabilityPgConnection`

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+@RequiredArgsConstructor
+@Service
+@Transactional
+public class DbMigrationGeneratorService {
+
+    private final HighAvailabilityPgConnectionFactory highAvailabilityPgConnectionFactory;
+    private final DataSource dataSource;
+    private final DbMigrationGenerator<ForeignKey> dbMigrationGenerator;
+
+    public ForeignKeyMigrationResponse addIndexesWithFkChecks(final ForeignKeyMigrationRequest fkMigrationRequest) {
```
Comment (mfvanek) [inline pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java:43]: The name of this method isn't clear. Try to find better name

File: pg-index-health-spring-boot-demo/build.gradle.kts [hunk 6]
```diff
@@ -10,6 +10,9 @@ plugins {
 
 dependencies {
     implementation(project(":db-migrations"))
+    implementation("io.github.mfvanek:pg-index-health-generator")
+    implementation(platform("org.springframework.boot:spring-boot-dependencies:3.3.4"))
```
[the remark this answers is not in the document]
Comment (marijarin) [reply pg-index-health-spring-boot-demo/build.gradle.kts:14]: The dependency in the line 13 is for using DbMigrationGenerator, ForeignKeyMigrationGenerator and GeneratingOptions classes from the library. And this dependency is used when migrations are generated in the module without Spring. Should I place this dependency 
```
implementation("io.github.mfvanek:pg-index-health-generator")
```
lower, where the similar dependencies are?  e.g. here https://github.com/mfvanek/pg-index-health-demo/blob/94ec8e607761286eaf5eada40eb9eafd85ec40ab/pg-index-health-spring-boot-demo/build.gradle.kts#L31

[the remark this answers is not in the document]
Comment (mfvanek) [reply pg-index-health-spring-boot-demo/build.gradle.kts:14]: You should remove
```
implementation(platform("org.springframework.boot:spring-boot-dependencies:3.3.4"))
implementation(platform("org.apache.httpcomponents.client5:httpclient5-parent:5.4"))
```

[the remark this answers is not in the document]
Comment (marijarin) [reply pg-index-health-spring-boot-demo/build.gradle.kts:14]: sorry for misunderstanding

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java (hunk 2 above)
Comment (marijarin) [reply pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java:63]: Could you, please, give me a hint what I am doing wrong? I tried to use JdbcTemplate here, but I faced a problem. All these "create" queries executed in one transaction, when executed in a loop. So I got an exception with a message "CREATE INDEX CONCURRENTLY cannot run inside a transaction block". I tried to put these statements in separate transactions, but failed.
Variants I implemented:
1. I used PreparedStatementCallback in order to close statement after executing. But I am not sure about the method inside lambda I chose.
This variant all statements are executed, but after all a rollback occured
image
2. I tried to enable autocommit mode using
```
  jdbcTemplate.getDataSource().getConnection().setAutoCommit(true);
```
but it didn't help.
3. I tried to add to statements "begin;" and "commit;" commands(with \n symbols at the end), but that didn't help. I got an error message about bad SQL grammar
4. I tried to use TransactionTemplate (with autowired PlatformTransactionManager ), but I also received exception with message  "CREATE INDEX CONCURRENTLY cannot run inside a transaction block"

image

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/config/MigrationGeneratorConfig.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+package io.github.mfvanek.pg.index.health.demo.config;
+
+import io.github.mfvanek.pg.generator.DbMigrationGenerator;
+import io.github.mfvanek.pg.generator.ForeignKeyMigrationGenerator;
+import io.github.mfvanek.pg.generator.GeneratingOptions;
+import io.github.mfvanek.pg.model.constraint.ForeignKey;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
```
Comment (mfvanek) [inline pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/config/MigrationGeneratorConfig.java:17]: Please mark as `@Configuration(proxyBeanMethods = false)`

File: config/spotbugs/exclude.xml [hunk 8]
```diff
@@ -15,6 +15,10 @@
         <Bug pattern="SIL_SQL_IN_LOOP"/>
         <Class name="io.github.mfvanek.pg.index.health.demo.utils.MigrationsGenerator"/>
     </Match>
+    <Match>
+        <Bug pattern="PRMC_POSSIBLY_REDUNDANT_METHOD_CALLS,SIL_SQL_IN_LOOP"/>
```
Comment (mfvanek) [inline config/spotbugs/exclude.xml:19]: Why `PRMC_POSSIBLY_REDUNDANT_METHOD_CALLS` was ignored?

**Review by mfvanek — CHANGES_REQUESTED** [review]

**Comment by mfvanek** [comment]
> I have another question. When running tests I noticed generated migrations be duplicated (1st and 2nd):
> 
> copied from logs
> 
> ```
> Generated migrations: [create index concurrently if not exists order_item_order_id_idx
>     on demo.order_item (order_id);, create index concurrently if not exists order_item_order_id_idx
>     on demo.order_item (order_id);, create index concurrently if not exists order_item_warehouse_id_idx
>     on demo.order_item (warehouse_id);, create index concurrently if not exists orders_buyer_id_idx
>     on demo.orders (buyer_id);, create index concurrently if not exists payment_order_id_without_nulls_idx
>     on demo.payment (order_id) where order_id is not null;]
> ```
> 
> Is it OK?

Yes, is it OK because we have duplicated foreign keys

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        final List<ForeignKey> keysBefore = getForeignKeysFromDb();
+        final List<String> migrations = generatedMigrations(keysBefore);
+        final List<ForeignKey> keysAfter = getForeignKeysFromDb();
+        if (!keysAfter.isEmpty()) {
+            throw new IllegalStateException("There should be no foreign keys not covered by the index");
+        }
+        return new ForeignKeyMigrationResponse(keysBefore, keysAfter, migrations);
+    }
+
+    public List<ForeignKey> getForeignKeysFromDb() {
```
[the remark this answers is not in the document]
Comment (marijarin) [reply pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java:48]: made package-private

File: config/spotbugs/exclude.xml (hunk 8 above)
Comment (marijarin) [reply config/spotbugs/exclude.xml:19]: I need to take foreign keys from DB twice to compare: before applying migrations and after that. SpotBugs produces warning beacause of this.

image

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationController.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    @PostMapping("/generate")
+    public ForeignKeyMigrationResponse generateMigrationsWithForeignKeysChecked() {
+        return dbMigrationGeneratorService.generateMigrationsWithForeignKeysChecked();
+    }
+
+    @ResponseStatus(HttpStatus.EXPECTATION_FAILED)
+    @org.springframework.web.bind.annotation.ExceptionHandler(IllegalStateException.class)
+    public MigrationError handleMigrationException(final IllegalStateException illegalStateException) {
+        return new MigrationError(HttpStatus.EXPECTATION_FAILED.value(), "Migrations failed: " + illegalStateException.getMessage());
```
[the remark this answers is not in the document]
Comment (marijarin) [reply pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationController.java:35]: Somethind went wrong and migrations wasn't applied to DB. What message will be more precise?

File: pg-index-health-spring-boot-demo/src/test/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationControllerMockTest.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+class DbMigrationControllerMockTest extends BasePgIndexHealthDemoSpringBootTest {
+
+    @MockBean
+    DbMigrationGeneratorService dbMigrationGeneratorService;
+
+    @Test
+    void returnsMigrationErrorWhenKeysAfterAreNotEmpty() {
+        final IllegalStateException illegalStateException = new IllegalStateException("There should be no foreign keys not covered by the index");
+        Mockito.when(dbMigrationGeneratorService.generateMigrationsWithForeignKeysChecked())
```
[the remark this answers is not in the document]
Comment (marijarin) [reply pg-index-health-spring-boot-demo/src/test/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationControllerMockTest.java:29]: In real situation there may be a case when somebody insert another row with foreign key concurrently. I need some time to implement something like that

File: pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mfvanek) [reply pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/controller/DbMigrationController.java:35]: Okay, let's leave this message.

**Review by mfvanek — APPROVED** [review]

**Pull request merged**